misc/fasta-sanitize.pl

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env perl
+#
+#    Copyright (C) 2020 Genome Research Ltd.
+#
+#    Author: James Bonfield <jkb@sanger.ac.uk>
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+
+# Usage: fasta-sanitize.pl [-t] file.in > file.out
+# Or via stdin, e.g. zcat file.in.gz | fasta-sanitize.pl [-t] > file.out
+#
+# Option -t ends names on a tab or newline, which permits spaces in names.
+# Otherwise names are considered to end on any whitespace.
+#
+# Also supports and autodetects fastq.
+
+# This tool sanitizes the reference names as per the SAM specification.
+# See SAM pull request https://github.com/samtools/hts-specs/pull/333
+
+# It is important that this is run prior to aligning.  This ensures the
+# SAM file contains @SQ lines which adhere to the specification, and that
+# VCF files produced from those SAM files also match the VCF specification.
+#
+# Furthermore, doing this early rather than via later with "samtools reheader"
+# means that the fasta file matches the SAM file.
+# Several samtools and bcftools sub-commands also require a fasta
+# reference, which must match the primary SQ lines, plus it helps CRAM
+# if using "view -T" (although it'll still work if using the
+# recommended practice of an MD5sum based ref cache).
+
+# The regexp permitted is:
+# [0-9A-Za-z!#$%&+./:;?@^_|~-][0-9A-Za-z!#$%&*+./:;=?@^_|~-]*
+
+use strict;
+
+my $re = qr/^[0-9A-Za-z!#$%&+.\/:;?@^_|~-][0-9A-Za-z!#$%&*+.\/:;=?@^_|~-]*$/;
+
+my $fastq = 0;
+my $in_qual = 0;
+my $seq_len = 0;
+
+my $name_whitespace = qr/^([>@])\s*(\S*)(.*)/;
+my $name_tab = qr/^([>@])\s*([^\t]*)(.*)/;
+
+my $name_re = $name_whitespace;
+if ($ARGV[0] eq "-t") {
+    pop(@ARGV);
```

Comment:
This doesn't work:
```
$ ./misc/fasta-sanitize.pl -t /tmp/test.fa
Can't open -t: No such file or directory at ./misc/fasta-sanitize.pl line 68.
```

I think `pop` should be `shift`.

Reply:
Oops, one of those last minute configuration tweaks that I clearly didn't test!

I just ditched it entirely as it was added due to user test data having spaces in names, but it later transpired it broke everything else so they'd already ran a sed script to remove all of those anyway.  An ill conceived attempt to work around something that's not an issue.